### PR TITLE
tb: refactor + support resuming from checkpoint after restart

### PR DESCRIPTION
**Pull request by benesch**
The headline feature here is support for resuming from the committed
checkpoint when tb restarts without rescanning the whole source
database. This doesn't guarantee that there won't be duplicates, but
Materialize should in theory be able to handle that deduplication.

This work includes a major refactor of tb as well. I apologize for
refactoring and changing behavior in one commit, but the whole of tb is
small enough that I hope it's not too hard to review. The new code
avoids mixing a bunch of static and non-static state together in one
large class, so it should be a good bit easier to follow.

Fix MaterializeInc/tb#1.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4077)

### CODE REVIEW

**Comment by benesch** [comment]
Oh, by the way, I backed out the fine-grained locking stuff. There is a synchronization bug in it as written (see Slack), and I don't think we really need to worry about holding a lock every second while we flush the files. At least, I doubt it will be an issue, and I think we can hold off on more complicated locking schemes until it is demonstrably problematic. There are probably five other ways tb fails to scale on databases with extremely high volume, and if we need to support those, we'll need to write some tb load tests and start fixing bottlenecks one-by-one.

File: misc/tb/src/TB.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+        ChangeWriter cw = new ChangeWriter(dir);
+        DebeziumEngine<RecordChangeEvent<SourceRecord>> engine = DebeziumEngine
+            .create(ChangeEventFormat.of(Connect.class))
+            .using(config)
+            .notifying(cw)
+            .build();
+
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(2);
+        executor.execute(engine);
```
Comment (damienhollis) [inline misc/tb/src/TB.java:128]: If DebeziumEngine crashes (e.g. can't connect to the server) tb keeps running. Maybe the DebeziumEngine should be wrapped so if it fails to start, System.exit(-1) can be called.

File: misc/tb/src/ChangeWriter.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        DataFileWriter dfw = schemaLogs.get(record.topic());
+        if (dfw == null) {
+            Schema schema = avroConverter.fromConnectSchema(record.valueSchema());
+            dfw = new DataFileWriter(new GenericDatumWriter(schema));
+            schemaLogs.put(record.topic(), dfw);
+
+            File file = this.dir.resolve(record.topic()).toFile();
+            if (file.exists()) {
+                logger.info("Appending to existing binlog {}", file);
```
Comment (quodlibetor) [inline misc/tb/src/ChangeWriter.java:82]: this seems like it should be trace level, as it seems like it may be emitted as often as once per row?

Comment (benesch) [reply misc/tb/src/ChangeWriter.java:82]: The containing block is only entered the first time you see a record for a topic, so should only get output once, I think.

File: misc/tb/src/TB.java (hunk 1 above)
Comment (benesch) [reply misc/tb/src/TB.java]: Whoops, yup. Thanks. Fixed.

**Comment by benesch** [comment]
I've reworked the `run` method so that it properly exits with a failing code if Debezium exits or if the flush task throws an exception. Please take another quick look!

File: misc/tb/src/TB.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        executor.scheduleWithFixedDelay(
+            new ExplodingRunnable(() -> cw.flushAll()),
+            0, 1, TimeUnit.SECONDS
+        );
+
+        // Run Debezium. If the task ever returns, something has gone wrong,
+        // so exit the process with a failing exit code. Debezium should have
+        // already logged an error message.
+        executor.submit(engine).get();
+        System.exit(1);
```
Comment (damienhollis) [inline misc/tb/src/TB.java:142]: Looks good!

Comment (benesch) [reply misc/tb/src/TB.java:142]: Thanks very much for taking a look!

**Pull request merged**